File: algo/src/main/java/com/hartwig/actin/algo/evaluation/priortumor/HasHistoryOfSecondMalignancyWithDOID.java
package com.hartwig.actin.algo.evaluation.priortumor;

import java.util.Set;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.datamodel.ImmutableEvaluation;
import com.hartwig.actin.algo.doid.DoidModel;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;
import com.hartwig.actin.clinical.datamodel.PriorSecondPrimary;

import org.jetbrains.annotations.NotNull;

public class HasHistoryOfSecondMalignancyWithDOID implements EvaluationFunction {

    @NotNull
    private final DoidModel doidModel;
    @NotNull
    private final String doidToMatch;

    HasHistoryOfSecondMalignancyWithDOID(@NotNull final DoidModel doidModel, @NotNull final String doidToMatch) {
        this.doidModel = doidModel;
        this.doidToMatch = doidToMatch;
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        boolean hasMatch = false;
        for (PriorSecondPrimary priorSecondPrimary : record.clinical().priorSecondPrimaries()) {
            if (isDoidMatch(priorSecondPrimary.doids(), doidToMatch)) {
                hasMatch = true;
            }
        }

        EvaluationResult result = hasMatch ? EvaluationResult.PASS : EvaluationResult.FAIL;
        ImmutableEvaluation.Builder builder = ImmutableEvaluation.builder().result(result);
        if (result == EvaluationResult.FAIL) {
            builder.addFailMessages("Patient has no history of second malignancy belonging to " + doidModel.term(doidToMatch));
        } else if (result == EvaluationResult.PASS) {
            builder.addPassMessages("Patient has history of second malignancy belonging to " + doidModel.term(doidToMatch));
        }

        return builder.build();
    }

    private boolean isDoidMatch(@NotNull Set<String> doids, @NotNull String doidToMatch) {
        for (String doid : doids) {
            if (doidModel.doidWithParents(doid).contains(doidToMatch)) {
                return true;
            }
        }

        return false;
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/priortumor/HasHistoryOfSecondMalignancyWithDoid.kt
package com.hartwig.actin.algo.evaluation.priortumor

import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.algo.evaluation.EvaluationLabels
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.doid.DoidModel

class HasHistoryOfSecondMalignancyWithDoid(
    private val doidModel: DoidModel,
    private val doidToMatch: String,
    private val labels: EvaluationLabels.PriorTumor
) : EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        val doidTerm = doidModel.resolveTermForDoid(doidToMatch)
        return if (record.priorPrimaries.flatMap { it.doids }.flatMap { doidModel.doidWithParents(it) }.contains(doidToMatch)) {
            EvaluationFactory.pass(labels.hasHistoryOfSecondMalignancyWithDoidPass(doidTerm.toString()))
        } else {
            EvaluationFactory.fail(labels.hasHistoryOfSecondMalignancyWithDoidFail(doidTerm.toString()))
        }
    }
}